sqlglot/optimizer/merge_subqueries.py

```diff
@@ -97,8 +97,11 @@ def merge_derived_tables(expression, leave_tables_isolated=False):
         for subquery in outer_scope.derived_tables:
             from_or_join = subquery.find_ancestor(exp.From, exp.Join)
             alias = subquery.alias_or_name
-            inner_scope = outer_scope.sources[alias]
-            if _mergeable(outer_scope, inner_scope, leave_tables_isolated, from_or_join):
+            inner_scope = outer_scope.sources.get(alias)
+
+            if inner_scope and _mergeable(
+                outer_scope, inner_scope, leave_tables_isolated, from_or_join
+            ):
```

Comment:
The invariant "every source has an alias" does not hold anymore, since you can have queries like:

```sql
SELECT * FROM (t1 JOIN t2)
```

There's a single scope here, since there's no alias attached to the nested source `(t1 JOIN t2)`.

Reply:
Update on this: the changes here were not needed after all because I "accidentally" solved the problem by not treating `(t1 join t2)` as a derived table.